rs/node_rewards/canister/src/lib.rs

```diff
@@ -5,5 +5,156 @@
 // 5. Simple Organization
 
+use crate::pb::v1::{
+    NodeMetrics as NodeMetricsProto, SubnetIdKey, SubnetMetricsKey, SubnetMetricsValue,
+};
+use candid::Principal;
+use ic_base_types::{PrincipalId, SubnetId};
+use ic_management_canister_types::NodeMetrics;
+use ic_stable_structures::storable::Bound;
+use ic_stable_structures::Storable;
+use prost::Message;
+use rewards_calculation::types::SubnetMetricsDailyKey;
+use std::borrow::Cow;
+
 pub mod canister;
+pub mod metrics;
+pub mod pb;
 pub mod storage;
+
+const HOUR_IN_SECONDS: u64 = 60 * 60;
+const DAY_IN_SECONDS: u64 = HOUR_IN_SECONDS * 24;
+const DAY_IN_NANOS: u64 = DAY_IN_SECONDS * 1_000_000_000;
+
+// Maximum sizes for the storable types chosen as result of test `max_bound_size`
+const MAX_BYTES_SUBNET_ID_STORED: u32 = 33;
+const MAX_BYTES_NODE_METRICS_STORED_KEY: u32 = 44;
+const PRINCIPAL_MAX_LENGTH_IN_BYTES: usize = 29;
+
+pub const MIN_PRINCIPAL_ID: PrincipalId = PrincipalId(Principal::from_slice(&[]));
+pub const MAX_PRINCIPAL_ID: PrincipalId = PrincipalId(Principal::from_slice(
+    &[0xFF_u8; PRINCIPAL_MAX_LENGTH_IN_BYTES],
+));
+
+impl From<NodeMetrics> for NodeMetricsProto {
+    fn from(metrics: NodeMetrics) -> Self {
+        NodeMetricsProto {
+            node_id: Some(metrics.node_id.into()),
+            num_blocks_proposed_total: metrics.num_blocks_proposed_total,
+            num_blocks_failed_total: metrics.num_block_failures_total,
+        }
+    }
+}
+
+impl Storable for SubnetIdKey {
+    fn to_bytes(&self) -> std::borrow::Cow<[u8]> {
+        let mut buf = Vec::with_capacity(self.encoded_len());
+        self.encode(&mut buf).unwrap();
+        Cow::Owned(self.encode_to_vec())
+    }
+
+    fn from_bytes(bytes: std::borrow::Cow<[u8]>) -> Self {
+        SubnetIdKey::decode(bytes.as_ref()).unwrap()
+    }
+
+    const BOUND: Bound = Bound::Bounded {
+        max_size: MAX_BYTES_SUBNET_ID_STORED,
+        is_fixed_size: false,
+    };
+}
+
+impl From<SubnetId> for SubnetIdKey {
+    fn from(subnet_id: SubnetId) -> Self {
+        Self {
+            subnet_id: Some(subnet_id.get()),
+        }
+    }
+}
+
+impl From<SubnetIdKey> for SubnetId {
+    fn from(subnet_id: SubnetIdKey) -> Self {
+        subnet_id.subnet_id.unwrap().into()
+    }
+}
+
+pub trait KeyRange {
+    fn min_key() -> Self;
+    fn max_key() -> Self;
+}
+
+impl From<SubnetMetricsKey> for SubnetMetricsDailyKey {
+    fn from(key: SubnetMetricsKey) -> Self {
+        Self {
+            day: key.timestamp_nanos.into(),
+            subnet_id: SubnetId::from(key.subnet_id.unwrap()),
+        }
+    }
+}
+
+impl Storable for SubnetMetricsKey {
+    fn to_bytes(&self) -> std::borrow::Cow<[u8]> {
+        Cow::Owned(self.encode_to_vec())
+    }
+
+    fn from_bytes(bytes: std::borrow::Cow<[u8]>) -> Self {
+        SubnetMetricsKey::decode(bytes.as_ref()).unwrap()
+    }
+
+    const BOUND: Bound = Bound::Bounded {
+        max_size: MAX_BYTES_NODE_METRICS_STORED_KEY,
```

Comment:
Previous comment was probably not clear -  but if you ever needed to add a new field to SubnetMetricsKey, then you might not be able to.   It might be a good idea to add some bytes beyond what's necessary now.